### PR TITLE
MODINREACH-159 React to Loan Record creation for INN-Reach Transactions 

### PR DESCRIPTION
**Pull request by oliinyko**
## Purpose
mod-inn-reach needs to be aware of and respond (when necessary) to the creation of loans in mod-circulation. The module should consume appropriate domain events for loans and respond accordingly when indicated.
https://issues.folio.org/browse/MODINREACH-159

## Approach
- Add Kafka event listener for circulation loan events
- Add tests

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java [hunk 1]
```diff
@@ -109,6 +110,19 @@ public ItemHoldCheckOutResponseDTO checkOutItemHoldItem(String itemBarcode, UUID
...
+  @Override
+  public void linkNewLoanToOpenTransaction(LoanDTO loan) {
+    var itemId = loan.getItemId();
+    var patronId = loan.getUserId();
+
+    transactionRepository.fetchOpenByFolioItemIdAndPatronId(itemId, patronId)
+      .ifPresent(transaction -> {
+        log.info("Linking open INN-Reach transaction {} with a new loan {}", transaction.getId(), loan.getId());
+        var hold = transaction.getHold();
+        hold.setFolioLoanId(loan.getId());
```
Comment (maryna-steshenko) [inline src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java]: You should also associate the due date of the loan.

**Review by maryna-steshenko — APPROVED** [review]

File: src/main/java/org/folio/innreach/domain/event/DomainEvent.java [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import lombok.ToString;
+
+@JsonInclude(JsonInclude.Include.NON_NULL)
+@Data
+@Builder
+@AllArgsConstructor
+@NoArgsConstructor
+@ToString(exclude = "data")
+public class DomainEvent<T> {
+  protected UUID recordId;
```
Comment (dmtkachenko) [inline src/main/java/org/folio/innreach/domain/event/DomainEvent.java:19]: if it corresponds to "id" field in the event then it's **event id** not record id

Comment (oliinyko) [reply src/main/java/org/folio/innreach/domain/event/DomainEvent.java:19]: it corresponds to the identifier of entity (loan) which we set manually from event's key
https://github.com/folio-org/mod-inn-reach/blob/601c92dc1f14e5a3e236f24d124ea710a1da1489/src/main/java/org/folio/innreach/domain/listener/KafkaCirculationEventListener.java#L53-L58

Comment (dmtkachenko) [reply src/main/java/org/folio/innreach/domain/event/DomainEvent.java:19]: ok

File: src/main/java/org/folio/innreach/domain/service/InnReachTransactionActionService.java [hunk 3]
```diff
@@ -13,4 +14,5 @@
 
   ItemHoldCheckOutResponseDTO checkOutItemHoldItem(String itemBarcode, UUID servicePointId);
 
+  void handleLoanCreation(LoanDTO loan);
```
Comment (dmtkachenko) [inline src/main/java/org/folio/innreach/domain/service/InnReachTransactionActionService.java]: associateNewLoanWithTransaction

**Review by dmtkachenko — APPROVED** [review]

**Pull request merged**